### PR TITLE
replace patchInterpreter function with PatchedInterpreter class

### PR DESCRIPTION
**Pull request by pcardune**
It's a lot easier to follow what's going on when we don't mutate globals.

Also upgrade js-interpreter-tyrant

### CODE REVIEW

**Comment by pcardune** [comment]
I'm currently running the full interpreter test suite locally with this change. Will update this diff with the saved results when it finishes (in about an hour).

File: apps/yarn.lock [hunk 1]
```diff
@@ -63,15 +63,9 @@
...
-  resolved "https://registry.yarnpkg.com/@code-dot-org/js-interpreter/-/js-interpreter-1.1.1.tgz#2a7ccbb6b1f31353445f82f777573d1c38e11d37"
-  dependencies:
-    yargs "^7.0.1"
-
-"@code-dot-org/js-interpreter-tyrant":
-  version "0.1.1"
-  resolved "https://registry.yarnpkg.com/@code-dot-org/js-interpreter-tyrant/-/js-interpreter-tyrant-0.1.1.tgz#34153bc7620748f07c687849741bbd7ba3d420a7"
+"@code-dot-org/js-interpreter-tyrant@0.2.2":
+  version "0.2.1"
+  resolved "https://registry.yarnpkg.com/@code-dot-org/js-interpreter-tyrant/-/js-interpreter-tyrant-0.2.1.tgz#bf7f20cc8689a1059cef8bd5a6e25a024eec9c84"
```
Comment (islemaster) [inline apps/yarn.lock]: Interesting... how did `0.2.2` resolve to `0.2.1`?

Comment (pcardune) [reply apps/yarn.lock]: weird. You know what, let me double check this. It might be that I had used yarn link to point to a local checkout of js-interpreter-tyrant and that somehow screwed up the resolution.

File: apps/package.json [hunk 2]
```diff
@@ -44,7 +44,7 @@
     "@code-dot-org/blockly": "1.6.18",
     "@code-dot-org/bramble": "0.1.14",
     "@code-dot-org/johnny-five": "0.10.10-cdo.0",
-    "@code-dot-org/js-interpreter-tyrant": "^0.1.1",
+    "@code-dot-org/js-interpreter-tyrant": "0.2.2",
```
Comment (islemaster) [inline apps/package.json]: Fine either way, but we're okay with

**Comment by pcardune** [comment]
Ok, fixed the issue with the yarn.lock file. This is ready to go.

**Comment by islemaster** [comment]
Can you summarize changes in the test results, since the diff is too large to view on GitHub?

**Comment by pcardune** [comment]
Here is the output of the diff:
```
New:
  es5: 0/22318
  es6: 0/19926
  es: 116/3627
  other: 0/781
Fixes:
  es5: 0/22318
  es6: 0/19926
  es: 0/3627
  other: 0/781
Regressions:
  es5: 21/22318
  es6: 0/19926
  es: 0/3627
  other: 0/781
```

All the es5 regressions are timeouts in "complex" regex tests. I'm not sure why they started timing out. I could try rerunning them with a longer timeout duration, but my assumption is that they are not that important to us.

**Comment by pcardune** [comment]
Update: I'm finding that if I run each of the regressed tests individually, then they pass. I think the issue is that when the tests run on circle, they run with one process per container. When I run them locally, I run it with one process per cpu (I have 8). There is probably some resource contention going on under the hood that makes something slower, causing the tests to time out. If I rerun the with a single process then it works, but takes *way* longer.

**Comment by pcardune** [comment]
Ok, after rerunning the regressed tests with a single process, they all pass. I've updated the saved test results accordingly.
